cli/emit.rs

```diff
@@ -642,66 +464,89 @@ pub fn emit(
 
-/// Check a module graph to determine if the graph contains anything that
-/// is required to be emitted to be valid. It determines what modules in the
-/// graph are emittable and for those that are emittable, if there is currently
-/// a valid emit in the cache.
-fn valid_emit(
+enum CheckHashResult {
+  Hash(u64),
+  NoFiles,
+}
+
+/// Gets a hash of the inputs for type checking. This can then
+/// be used to tell
+fn get_check_hash(
   graph_data: &GraphData,
-  cache: &dyn EmitCache,
-  ts_config: &TsConfig,
-  reload: bool,
-  reload_exclusions: &HashSet<ModuleSpecifier>,
-) -> bool {
-  let config_bytes = ts_config.as_bytes();
-  let check_js = ts_config.get_check_js();
-  for (specifier, module_entry) in graph_data.entries() {
+  options: &CheckOptions,
+) -> CheckHashResult {
+  // twox hash is insecure, but fast so it works for our purposes
+  use std::hash::Hasher;
+  use twox_hash::XxHash64;
+
+  let mut hasher = XxHash64::default();
+  hasher.write_u8(match options.type_check_mode {
+    TypeCheckMode::All => 0,
+    TypeCheckMode::Local => 1,
+    TypeCheckMode::None => 2,
+  });
+  hasher.write(&options.ts_config.as_bytes());
+
+  let check_js = options.ts_config.get_check_js();
+  let mut sorted_entries = graph_data.entries().collect::<Vec<_>>();
+  sorted_entries.sort_by_key(|(s, _)| s.as_str()); // make it deterministic
+  let mut has_file = false;
+  let mut has_file_to_type_check = false;
+  for (specifier, module_entry) in sorted_entries {
     if let ModuleEntry::Module {
       code,
       media_type,
       ts_check,
       ..
     } = module_entry
     {
+      if *ts_check {
+        has_file_to_type_check = true;
+      }
+
       match media_type {
         MediaType::TypeScript
+        | MediaType::Dts
+        | MediaType::Dmts
+        | MediaType::Dcts
         | MediaType::Mts
         | MediaType::Cts
-        | MediaType::Tsx
-        | MediaType::Jsx => {}
-        MediaType::JavaScript | MediaType::Mjs | MediaType::Cjs => {
+        | MediaType::Tsx => {
+          has_file = true;
+          has_file_to_type_check = true;
+        }
+        MediaType::JavaScript
+        | MediaType::Mjs
+        | MediaType::Cjs
+        | MediaType::Jsx => {
+          has_file = true;
           if !check_js && !ts_check {
             continue;
           }
         }
         MediaType::Json
         | MediaType::TsBuildInfo
         | MediaType::SourceMap
-        | MediaType::Dts
-        | MediaType::Dmts
-        | MediaType::Dcts
         | MediaType::Wasm
         | MediaType::Unknown => continue,
       }
-      if reload && !reload_exclusions.contains(specifier) {
-        return false;
-      }
-      if let Some(source_hash) = cache.get_source_hash(specifier) {
-        if source_hash != get_version(code.as_bytes(), &config_bytes) {
-          return false;
-        }
-      } else {
-        return false;
-      }
+      hasher.write(specifier.as_str().as_bytes());
+      hasher.write(code.as_bytes());
```

Comment:
Would it be better to use precomputed hashes of the code here instead of the code itself?

Reply:
Maybe. I wouldn't be surprised if retrieving and storing those hashes is slower than just doing the hash though. xxhash is very fast.